SciLean implementation of GMM eval

lecopivo opened the pull request:
Implementation of GMM eval with SciLean

athas commented:
I am highly excited for this PR. Gradbench should maximise the number of exotic languages used for tools!

lecopivo commented:
And I'm really glad you are working on gradbench. It really helps me to be more focused on developing AD in SciLean

samestep commented:
Really awesome work, hopefully we can get this merged soon! Just to clarify, the only reason I haven't already merged this is that (last I checked) the multi-platform Docker build doesn't work yet. See also https://github.com/gradbench/gradbench/pull/233#issuecomment-2646636367 which may be a similar situation.

lecopivo commented:
Yeah this is a problem :( unfortunately right now I have trouble building SciLean on anything else but Linux/Ubuntu. It might take me some time to fix this

lecopivo commented:
It builds and runs but I updated SciLean to a newer version where I switched to row-major matrices by default and I'm getting incorrect results for the gradients now :)

samestep commented:
@lecopivo I haven't looked too closely yet, but it seems like the multi-platform build is failing at the `lake exe cache get` step; maybe this could be solved by building a `mathlib4` base image like what we've been doing before this PR?

lecopivo commented:
I think the `lake exe cache get` fails just because of some network issues. I would try to run it again and let's see if it still a problem. I have experienced this step failing during CI few times before.

lecopivo commented:
Anyway, it is probably a good idea to build the `mathlib4` base image as you did before. I just need to learn how to use it and how to update to new version of mathlib when the need comes.

Is there a way to fool-proof this? When I run `lake update scilean` I get the newest version of scilean which might use new version of lean and new version of mathlib. Can the `mathlib4` docker image be of the exact revision given in [`lake-manifest.json`](https://github.com/gradbench/gradbench/blob/e4da4c3280269e48b59cc1d14da3e1cb9113ea34/tools/scilean/lake-manifest.json#L61) ?

samestep commented:
> I looked into the log a bit more and I'm a bit confused. I seems to run `lake exe cache get` and `lake build` multiple times. The log contains lines like
> 
> ```
> #32 4214.1 ✔ [3290/5437] Built Mathlib.Algebra.Module.Submodule.Pointwise:c.o
> ```
> 
> which is a clear indication that it is building thus got passed `lake exe cache get` but then the error is while running `lake exe cache get`.
> 
> Is `Build / tool (scilean) (pull_request) ` and its step `Build tool Docker image` building multiple versions for different platforms at the same time?

Yes, sorry for the confusion; the `Build / tool` job passes the `--cross` flag to `gradbench repo build-tool`:

https://github.com/gradbench/gradbench/blob/e4da4c3280269e48b59cc1d14da3e1cb9113ea34/.github/workflows/build.yml#L233

And this flag adds the argument `--platform linux/amd64,linux/arm64` to `docker build`:

https://github.com/gradbench/gradbench/blob/e4da4c3280269e48b59cc1d14da3e1cb9113ea34/crates/gradbench/src/main.rs#L290-L292

This causes Docker to run the two builds in parallel, interleaving their logs but denoting them like this:

```
#15 [linux/amd64  7/15] COPY tools/scilean/lake-manifest.json .
#15 DONE 0.0s

#16 [linux/amd64  8/15] COPY tools/scilean/lean-toolchain .
#16 DONE 0.0s

#17 [linux/amd64  9/15] RUN lake exe cache get      # Fetch the cache for dependencies
#17 0.713 info: downloading https://github.com/leanprover/lean4/releases/download/v4.16.0/lean-4.16.0-linux.tar.zst
#17 2.503 info: installing /root/.elan/toolchains/leanprover--lean4---v4.16.0
#17 7.445 info: scilean: cloning https://github.com/lecopivo/SciLean
#17 7.445 info: scilean: checking out revision 'aa2839689dbd48d433bcf4ad192062e818ec0b26'
#17 8.189 info: leanblas: cloning https://github.com/lecopivo/LeanBLAS
#17 8.189 info: leanblas: checking out revision '9d2c7a8ced83d21d1adbcaeaa9da894034772de8'
#17 ...

#9 [linux/arm64  2/15] RUN apt-get update && apt-get install -y curl git jq build-essential libopenblas-dev
#9 11.45 Reading package lists...
#9 60.46 Reading package lists...
#9 ...
```



> Anyway, it is probably a good idea to build the `mathlib4` base image as you did before. I just need to learn how to use it and how to update to new version of mathlib when the need comes.
> 
> Is there a way to fool-proof this? When I run `lake update scilean` I get the newest version of scilean which might use new version of lean and new version of mathlib. Can the `mathlib4` docker image be of the exact revision given in [`lake-manifest.json`](https://github.com/gradbench/gradbench/blob/e4da4c3280269e48b59cc1d14da3e1cb9113ea34/tools/scilean/lake-manifest.json#L61) ?

Yes, that's what the current `mathlib4` Docker image does:

https://github.com/gradbench/gradbench/blob/e4da4c3280269e48b59cc1d14da3e1cb9113ea34/docker/mathlib4/lakefile.lean#L6

So we'd just need to modify it to use a different `mathlib4` version. But this isn't super great because it's very manual. I'm trying to figure out if it'd be possible to do a more automatic/general solution, e.g. to make a separate `gradbench/mathlib4-docker` repo or something which automatically builds Docker images for mathlib4 commits. But I'm not sure the best way to structure that...

lecopivo commented:
I made a PR #250 with a python script that updates lakefile.lean, lean-toolchain and lake-manifest.json in the mathlib docker image to match the ones in tools/scilean.

Not sure when and where to run this automatically, but even if you are supposed to run it manually it will be a lot easier.

samestep commented:
OK, in #268 I built and pushed a new mathlib image, so you should be able to base the SciLean Docker build from it like this:

```dockerfile
FROM ghcr.io/gradbench/mathlib4:93904a0d036e5321062ae5e705e8a893da24d81b
```

@lecopivo could you try that and see if it fixes the CI build for this PR? I'm a bit worried that it won't, because [the logs](https://github.com/gradbench/gradbench/actions/runs/13316955541/job/37231631956?pr=224) seem to indicate that this is an issue with the way we're trying to build BLAS for ARM:

```
#32 6302.7 trace: .> /root/.elan/toolchains/leanprover--lean4---v4.16.0/bin/clang -shared -o ././.lake/packages/scilean/.lake/build/lib/libscileanc.so -Wl,--whole-archive ././.lake/packages/scilean/.lake/build/lib/libscileanc.a -Wl,--no-whole-archive -L/usr/lib/x86_64-linux-gnu/ -lblas -lm --sysroot /root/.elan/toolchains/leanprover--lean4---v4.16.0 -L /root/.elan/toolchains/leanprover--lean4---v4.16.0/lib -L /root/.elan/toolchains/leanprover--lean4---v4.16.0/lib/glibc /root/.elan/toolchains/leanprover--lean4---v4.16.0/lib/glibc/libc_nonshared.a /root/.elan/toolchains/leanprover--lean4---v4.16.0/lib/glibc/libpthread_nonshared.a -Wl,--as-needed -Wl,-Bstatic -lgmp -lunwind -luv -Wl,-Bdynamic -Wl,--no-as-needed -fuse-ld=lld -L /root/.elan/toolchains/leanprover--lean4---v4.16.0/lib/lean -Wl,-Bstatic -lc++ -lc++abi -Wl,-Bdynamic -Wl,--as-needed -lLake_shared -Wl,--no-as-needed -Wl,-Bsymbolic -Wl,--as-needed -lgmp -luv -lpthread -ldl -lrt -Wl,--no-as-needed -lm -ldl -pthread
#32 6302.7 info: stderr:
#32 6302.7 ld.lld: error: unable to find library -lblas
#32 6302.7 clang: error: linker command failed with exit code 1 (use -v to see invocation)
#32 6302.7 error: external command '/root/.elan/toolchains/leanprover--lean4---v4.16.0/bin/clang' exited with code 1
```

lecopivo commented:
Yes that is definitely issue with BLAS. In general, I'm having really hard time getting BLAS working properly now. I will try to figure out how to build SciLean for `ubuntu-latest`, `macos-latest` and `macos-13` first. Right now, I can't even build SciLean on `ubuntu-latest` for some reason, locally it works no problem :(

lecopivo commented:
I'm currently writing a grant proposal and I have to completely focus on that for the next two weeks. I was getting some really mysterious errors when compiling SciLean on arm and more investigation needs to be done. I don't think you can help right now but I appreciate the offer!

I might need some graphs for the grant so I might actually do it but can't promise anything right now :)

samestep commented:
@lecopivo I merged #318 which makes an exception for SciLean so CI doesn't try to build it on ARM. I think that makes this PR ready to merge! The only thing I had to fix in 210f92568e381149f70e0dd1d2c9b45bcf826f05 is because of the changes made in #282. Technically the current SciLean implementation doesn't do what it is supposed to do with the `min_runs` and `min_seconds` inputs it is given, but I think that can be done in a followup PR; the only problem this should cause in the meantime is noisier performance data.

@athas just to check, are you OK with merging this PR before it correctly handles `min_runs` and `min_seconds`, or would you prefer to wait until it satisfies those? Also, note that this PR doubles the amount of time it takes to build the x86 Docker image for the SciLean tool in CI, which means it will now take about twice as long before the `run` job starts.

samestep reviewed (approved):
@lecopivo I merged #318 which makes an exception for SciLean so CI doesn't try to build it on ARM. I think that makes this PR ready to merge! The only thing I had to fix in 210f92568e381149f70e0dd1d2c9b45bcf826f05 is because of the changes made in #282. Technically the current SciLean implementation doesn't do what it is supposed to do with the `min_runs` and `min_seconds` inputs it is given, but I think that can be done in a followup PR; the only problem this should cause in the meantime is noisier performance data.

@athas just to check, are you OK with merging this PR before it correctly handles `min_runs` and `min_seconds`, or would you prefer to wait until it satisfies those? Also, note that this PR doubles the amount of time it takes to build the x86 Docker image for the SciLean tool in CI, which means it will now take about twice as long before the `run` job starts.

samestep commented:
Hmm, seems like there's some sort of bug in the logic I've been using to check for permissions:

https://github.com/gradbench/gradbench/blob/39ac6179e97923db4328bd55e3074c697c8d9f01/.github/workflows/build.yml#L270-L280

That's unrelated to this PR, so I don't think it affects whether we should merge it; I'll just open a separate GitHub issue.

athas commented:
I am fine with this PR.

samestep commented:
@lecopivo it looks like CI is passing. Should I merge this?
